### PR TITLE
Start Coding Style section

### PR DESCRIPTION
**Pull request by mophahr**
Here's my draft for a Coding Style section. I have moved the parts from the Roles section as the guidelines are more general and I have cleaned them up a bit and updated them where necessary.

### CODE REVIEW

File: coding_style/README.adoc [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+* Use a single space separating the template markers from the variable name inside all Jinja2 template points.
+  For instance, always write it as `{{ variable_name_here }}`.
+  The same goes if the value is an expression. `{{ variable_name | default('hiya, doc') }}`
+* When naming files, use the `.yml` extension and _not_ `.yaml`.
+  `.yml` is what `ansible-galaxy init` does when creating a new role template.
+* Use double quotes for YAML strings with the exception of Jinja2 strings which will use single quotes.
+* Do not use quotes unless you have to, especially for short module-keyword-like strings like `present`, `absent`, etc.
+  But do use quotes for user-side strings such as descriptions, names, and messages.
+
+== Python Guidelines
```
Comment (ericzolf) [inline coding_style/README.adoc:112]: I would move the python guidelines to the plugin section.

File: coding_style/README.adoc [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+* Review Ansible guidelines for https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_best_practices.html[modules] and https://docs.ansible.com/ansible/latest/dev_guide/index.html[development].
+* Use https://pep8.org/[PEP8].
+* File headers and functions should have comments for their intent.
+
+== Ansible Guidelines
+
+* Ensure that all tasks are idempotent.
+* https://github.com/ansible/ansible/issues/10374[Ansible variables use lazy evaluation.]
+* Prefix all tags within a role with the role name or, alternatively, a "unique enough" but descriptive prefix.
```
Comment (ericzolf) [inline coding_style/README.adoc:122]: That belongs in the roles section.

File: coding_style/README.adoc [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  Even better, use a dedicated module, if it exists.
+  If not, see the <<check-mode-and-idempotency-issues,section>> about idempotency and check mode and make sure that your task is idempotent and supports check mode properly;
+  your task will likely need options such as `changed_when:` and maybe `check_mode:`).
+* Anytime `command` or `shell` modules are used, add a comment in the code with justification to help with future maintenance.
+* Use the `| bool` filter when using bare variables (expressions consisting of just one variable reference without any operator) in `when`.
+* Do not use `meta: end_play`.
++
+[%collapsible]
+====
+Explanation:: It aborts the whole play instead of a given host (with multiple hosts in the inventory).
```
Comment (ericzolf) [inline coding_style/README.adoc:133]: It's the rationale, not the explanation.

File: coding_style/README.adoc [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+  If absolutely necessary, consider using `meta: end_host`.
+====
++
+* Task names can be made dynamic by using variables (wrapped in Jinja2 templates), this helps with reading the logs.
+* Do not use variables (wrapped in Jinja2 templates) for play names, variables don't get expanded properly there.
```
Comment (ericzolf) [inline coding_style/README.adoc:138]: Also not in loops

File: coding_style/README.adoc [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+Modifying config files with it can cause the Ansible code to become arcane and difficult to read, especially for someone not familiar with the file in question.
+Try editing files directly using other built-in modules (e.g. `ini_file`, `blockinfile`, `xml`), or reading and parsing.
+If you are modifying more than a tiny number of lines or in a manner more than trivially complex, try leveraging the `template` module, instead.
+This will allow the entire structure of the file to be seen by later users and maintainers.
+The use of `lineinfile` should include a comment with justification.
+====
++
+* Limit use of the `copy` module to copying remote files and to uploading binary blobs.
+  For all other file pushes, use the `template` module. Even if there is nothing in the file that is being templated at the current moment, having the file handled by the `template` module now makes adding that functionality much simpler than if the file is initially handled by the `copy` and then needs to be moved before it can be edited.
+# TODO: I am not convinced about the following rule:
```
Comment (ericzolf) [inline coding_style/README.adoc:196]: We should separate the reshuffling from changing rules, there are also a few other places which don't convince me, but let's have this discussion in separate PRs. Create an issue to start the discussion.

File: coding_style/README.adoc [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+* Keep filenames and templates as close to the name on the destination system as possible.
++
+[%collapsible]
+====
+Rationale:: This will help with both editor highlighting as well as identifying source and destination versions of the file at a glance.
+Avoid duplicating the remote full path in the role directory, however, as that creates unnecessary depth in the file tree for the role.
+Grouping sets of similar files into a subdirectory of `templates` is allowable, but avoid unnecessary depth to the hierarchy.
+====
++
+* Use `{{ role_path }}/subdir/` as the filename prefix when including files if the name has a variable in it.
```
Comment (ericzolf) [inline coding_style/README.adoc:208]: That's specific to roles and belongs in this section.

**Review by ericzolf — CHANGES_REQUESTED** [review]

File: coding_style/README.adoc [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+example_list:
+- example_element_1
+- example_element_2
+- example_element_3
+- example_element_4
+----
+====
++
+* Split long Jinja2 expressions into multiple
+lines.
```
[the remark this answers is not in the document]
Comment (mophahr) [reply coding_style/README.adoc:58]: fixed

File: coding_style/README.adoc [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+* If the `when:` condition results in a line that is too long, and is an `and`
+expression, then break it into a list of conditions.
```
[the remark this answers is not in the document]
Comment (mophahr) [reply coding_style/README.adoc:60]: fixed

File: coding_style/README.adoc [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
++
+[%collapsible]
+====
+Rationale:: Ansible will `and` thetogether (https://docs.ansible.coansible/latest/user_guidplaybooks_conditionalhtml#the-when-statement[Ansible UseGuide » Conditionals]).
```
[the remark this answers is not in the document]
Comment (mophahr) [reply coding_style/README.adoc:64]: I added a few

File: coding_style/README.adoc [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+Multiple conditions that all need tbe true (a logical `and`) can alsbe specified as a list, but bewarof bare variables in `when:`.
+====
++
+* All roles need to, minimally, pass a basic ansible-playbook syntax check run
+* All task arguments should be spelled out in YAML style and not use `key=value` type of arguments
```
[the remark this answers is not in the document]
Comment (mophahr) [reply coding_style/README.adoc:69]: fixed

File: coding_style/README.adoc [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
++
+* Avoid the use of `True` and `False` for boolean values in playbooks.
++
+[%collapsible]
+====
+Explanation:: These values are sometimes used because they are the words Python uses.
+  However, they are improper YAML and will be treated as either strings or as booleans but generating a warning depending on the particular YAML implementation.
+====
++
+* Do not use the Ansible-specific `yes` and `no` as boolean values in YAML as these are completely custom extensions used by Ansible and are not part of the YAML spec.
```
[the remark this answers is not in the document]
Comment (mophahr) [reply coding_style/README.adoc:99]: addressed it, please er-check

File: coding_style/README.adoc (hunk 2 above)
Comment (mophahr) [reply coding_style/README.adoc:122]: moved.

File: coding_style/README.adoc (hunk 4 above)
Comment (mophahr) [reply coding_style/README.adoc:138]: I don't understand this comment

File: coding_style/README.adoc (hunk 5 above)
Comment (mophahr) [reply coding_style/README.adoc:196]: Agreed. 

I removed the `TODO` line and opened https://github.com/redhat-cop/automation-good-practices/issues/16

File: coding_style/README.adoc (hunk 6 above)
Comment (mophahr) [reply coding_style/README.adoc:208]: moved.

**Comment by mophahr** [comment]
@ericzolf I have addressed all your comments. Could you please have another look?

File: coding_style/README.adoc (hunk 4 above)
Comment (ericzolf) [reply coding_style/README.adoc]: Loop variables (by default "item") also don't get properly expanded in looped tasks' name and hence are not to be used there.

**Review by ericzolf — APPROVED** [review]

**Comment by ericzolf** [comment]
You introduced a conflict which you need to resolve before the PR can be merged but else, it looks good.

**Comment by jscar-hawk** [comment]
At the 2 week automation-cop review,
Moritz reworked the role section.

There are additional concerns, but we have made other issues to address them.
This Pull Request is Approved!

**Pull request merged**